### PR TITLE
Support for multiple output devices

### PR DESCRIPTION
**Pull request by rshym**
Provides support for multiple output devices.

### CODE REVIEW

File: device.c [hunk 1]
```diff
@@ -80,7 +83,7 @@ int adev_open(const hw_module_t* module, const char* name, hw_device_t** device)
 
     if (ref_counter != 0) {
         *device = &xadev->hw_device.common;
-        ALOGD("Return existing device:%p", *device);
+        LOG_FN_PARAMETERS("Return existing device:%p", *device);
```
Comment (andr2000) [inline device.c]: Don't think this and below changes of ALOG->LOG_FN are relevant

File: device.c [hunk 2]
```diff
@@ -316,25 +324,42 @@ int adev_open_output_stream(struct audio_hw_device *dev,
...
         pthread_mutex_unlock(&xdev->lock);
         return -EINVAL;
     }
 
-    /* TODO define card_id and dev_id for stream, using provided parameters:
-       devices, flags, address, source */
+    /* For now we have one card, so card id is always constant */
+    /* Get device id. We expect (same as parser in car audio service)
+     * that address has format "busX_%s" */
+    if (sscanf(address, "bus%u", &dev_id) != 1) {
```
Comment (andr2000) [inline device.c]: if address is a malformed, e.g. it is bus7bla, then sscanf above will return 7 as dev_id, but ignore "bla".
Do you need to check that address is exactly "bus7" in the example above, not "bus7bla?

File: device.c [hunk 3]
```diff
@@ -316,25 +324,42 @@ int adev_open_output_stream(struct audio_hw_device *dev,
...
+        ALOGE("Address has not supported format.");
```
Comment (andr2000) [inline device.c]: I would change wording here to be more... precise?

File: device.c [hunk 4]
```diff
@@ -316,25 +324,42 @@ int adev_open_output_stream(struct audio_hw_device *dev,
...
+            if (*stream_out != NULL) {
+                xdev->xout_streams[dev_id] = (x_stream_out_t*)(*stream_out);
+                if (xdev->master_is_muted) {
+                    out_set_mute(xdev->xout_streams[dev_id], true);
+                }
+                LOG_FN_PARAMETERS("Created stream_out:%p", *stream_out);
+            }
+            /* if out_create failed then 'res' has 'failed' value and it will be reported up */
+        } else {
+            /* May be we have to return existing stream here? */
```
Comment (andr2000) [inline device.c]: Please consider adding "FIXME" tag

File: device.c [hunk 5]
```diff
@@ -138,6 +141,8 @@ int adev_open(const hw_module_t* module, const char* name, hw_device_t** device)
 
 int adev_close(hw_device_t *device)
 {
+    uint16_t i = 0;
```
Comment (andr2000) [inline device.c]: why not int?

File: device.c [hunk 6]
```diff
@@ -343,17 +368,22 @@ int adev_open_output_stream(struct audio_hw_device *dev,
 void adev_close_output_stream(struct audio_hw_device *dev, struct audio_stream_out* stream_out)
 {
     x_audio_device_t *xdev = (x_audio_device_t*)dev;
+    uint16_t i = 0;
```
Comment (andr2000) [inline device.c]: ditto

File: device.c [hunk 7]
```diff
@@ -158,14 +163,17 @@ int adev_close(hw_device_t *device)
         if (xadev->xin_stream != NULL) {
             in_destroy(xadev->xin_stream);
         }
-        if (xadev->xout_stream != NULL) {
-            out_destroy(xadev->xout_stream);
+        for (i=0; i<NUMBER_OF_DEVICES_OUT; i++) {
```
Comment (andr2000) [inline device.c]: Coding style: I didn't get if you use spaces around "=" or no? Here you omit spaces and when you use "if (XX ==" you put one

File: device.c [hunk 8]
```diff
@@ -444,6 +474,7 @@ int adev_get_microphones(const struct audio_hw_device *dev,
 int adev_dump(const struct audio_hw_device *dev, int fd)
 {
     x_audio_device_t *xdev = (x_audio_device_t*)dev;
+    uint16_t i = 0;
```
Comment (andr2000) [inline device.c]: ditto

File: device.c [hunk 9]
```diff
@@ -463,15 +496,18 @@ int adev_dump(const struct audio_hw_device *dev, int fd)
 int adev_set_master_mute(struct audio_hw_device *dev, bool mute)
 {
     x_audio_device_t *xdev = (x_audio_device_t*)dev;
+    uint16_t i = 0;
```
Comment (andr2000) [inline device.c]: ditto

File: stream_in.c [hunk 10]
```diff
@@ -264,7 +264,7 @@ int in_get_capture_position(const struct audio_stream_in *stream,
         *time = xin->last_time;
     }
     /* this trace generates lots of messages */
-    /* ALOGD("igcp: %" PRIu64 ", %" PRIu64, xin->read_frames, xin->last_time); */
+    /* LOG_FN_PARAMETERS("igcp: %" PRIu64 ", %" PRIu64, xin->read_frames, xin->last_time); */
```
Comment (andr2000) [inline stream_in.c]: Please remove dead code

**Review by andr2000 — CHANGES_REQUESTED** [review]

File: device.c (hunk 1 above)
Comment (rshym) [reply device.c]: This change is removed from pull request, and will be submitted separately.

File: device.c (hunk 2 above)
Comment (rshym) [reply device.c]: additional text is expected after number, so comment was expanded to provide additional info.

File: device.c (hunk 3 above)
Comment (rshym) [reply device.c]: Done. Explanation was added to message.

File: device.c (hunk 4 above)
Comment (rshym) [reply device.c]: This comment is removed. It was my internal note.

File: device.c (hunk 5 above)
Comment (rshym) [reply device.c]: Fixed.

File: device.c (hunk 6 above)
Comment (rshym) [reply device.c]: Fixed.

File: device.c (hunk 7 above)
Comment (rshym) [reply device.c]: Fixed. Spaces are used inside 'for'.

File: device.c (hunk 8 above)
Comment (rshym) [reply device.c]: Fixed.

File: device.c (hunk 9 above)
Comment (rshym) [reply device.c]: Fixed.

File: stream_in.c (hunk 10 above)
Comment (rshym) [reply stream_in.c]: Thanks for pointing. This fix will be issued as separate patch in separate pull request, as far as it has no relation to output devices.

File: device.c [hunk 11]
```diff
@@ -316,25 +323,43 @@ int adev_open_output_stream(struct audio_hw_device *dev,
...
 
-    /* TODO define card_id and dev_id for stream, using provided parameters:
-       devices, flags, address, source */
+    /* For now we have one card, so card id is always constant */
+    /* Get device id. We expect (same as parser in car audio service)
+     * that address has format "bus%d_%s".
+     * In other words, we expect that bus address starts with 'bus',
+     * followed by bus number, which is followed by any voluntary description. */
+    if (sscanf(address, "bus%u", &dev_id) != 1) {
+        ALOGE("Address has not supported format. Expect 'bus%d_%s': 'bus' word, bus number, '_', description.");
```
Comment (andr2000) [inline device.c]: I would say: "Address format is not supported...."

**Review by andr2000 — COMMENTED** [review]
I would say that the commit message should be more descriptive because as it is now it doesn't bring any value for the reader

**Comment by andr2000** [comment]
`Reviewed-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>`

**Comment by rshym** [comment]
Remove dependency on pull request 114 in meta-xt-prod-devel.

**Comment by arminn** [comment]
Reviewed-by: Andrii Chepurnyi <andrii_chepurnyi@epam.com>

**Review by andr2000 — APPROVED** [review]

**Comment by rshym** [comment]
Added "Reviewed-by:".

**Pull request merged**